File: 670-maximum-swap/maximum-swap.go
func maximumSwap(num int) int {
    ans:=0
    digits:=[]int{}
    for num>0{
        digits = append(digits,num%10)
        num/=10
    }  
    n:=len(digits)
    for i:=0;i<n/2;i++{
        digits[i],digits[n-i-1] = digits[n-1-i],digits[i]
    }  
    greater:=make([]int,n)
    greater[n-1] = n-1
    for i:=n-2;i>=0;i--{ 
        if digits[i]>digits[greater[i+1]]{
            greater[i] = i
        }else{
            greater[i] = greater[i+1]
        }
    }

    for i:=0;i<n;i++{
        if digits[i]<digits[greater[i]]{
            digits[i],digits[greater[i]] = digits[greater[i]],digits[i]
            break
        }
    }

    for i:=0;i<n;i++{
        ans = ans*10 + digits[i]
    }
    return ans
}

File: 670-maximum-swap/maximum-swap.cpp
class Solution {
public:
    int maximumSwap(int num) {
        vector<int>digits;
        while(num){
            digits.push_back(num%10);
            num/=10;
        }        
        reverse(digits.begin(),digits.end());
        int n = digits.size();

        vector<int>greater(n);
        greater[n-1] = n-1;
        for(int i=n-2;i>=0;i--){
            if(digits[i]>digits[greater[i+1]]){
                greater[i]=i;
            }else{
                greater[i]=greater[i+1];
            }
        }

        for(int i=0;i<n;i++){
            if(digits[i]<digits[greater[i]]){
                swap(digits[i],digits[greater[i]]);
                break;
            }
        }

        int ans = 0;
        for(int i=0;i<n;i++){
            ans = ans*10 + digits[i];
        }
        return ans;
    }
};